include/hydrogen/blas/gpu/CombineImpl.hpp

```diff
@@ -0,0 +1,213 @@
+ *  @tparam FunctorT (Inferred) Type of functor. Must be equivalent to
+ *                              `T(S const&, T const&)`.
+ *
+ *  @param num_entries The number of entries to which the functor is
+ *                     applied.
+ *  @param A The source data buffer.
+ *  @param stride_A The stride between elements of A in terms of
+ *                  elements of type S.
+ *  @param B The target data buffer.
+ *  @param stride_B The stride between elements of B in terms of
+ *                  elements of type T.
+ *  @param func The functor to apply. Must be device-invocable.
+ */
+template <typename S, typename T, typename SizeT, typename FunctorT>
+void CombineImpl(
+    SizeT size,
+    S const* A, SizeT stride_A,
+    T * B, SizeT stride_B,
+    FunctorT func,
+    SyncInfo<Device::GPU> const& sync_info)
+{
+    if (size == 0)
+        return;
+
+    constexpr size_t threads_per_block = 256ULL;
+    auto blocks = (size + threads_per_block - 1) / threads_per_block;
+    gpu::LaunchKernel(
+        kernel::combine_1d_kernel_naive<
+            NativeGPUType<S>, NativeGPUType<T>, SizeT, FunctorT>,
+        blocks, threads_per_block, 0, sync_info,
+        size,
+        AsNativeGPUType(A), stride_A,
+        AsNativeGPUType(B), stride_B,
+        func);
+}
+
+/** @brief Apply a functor to 2-D column-major matrix buffers.
+ *
+ *  This can be applied to a row-major matrix by logically transposing
+ *  the matrix.
+ *
+ *  @warning Calling this function is only valid in device code.
+ *
+ *  @tparam S (Inferred) Type of source buffer.
+ *  @tparam T (Inferred) Type of target buffer.
+ *  @tparam SizeT (Inferred) Type of integer used to express sizes.
+ *  @tparam FunctorT (Inferred) Type of functor. Must be equivalent to
+ *                              `T(S const&, T const&)`.
+ *
+ *  @param m The number of rows in A/B.
+ *  @param n The number of columns in A/B. Columns must be contiguous
+ *           in memory.
+ *  @param A The source matrix buffer.
+ *  @param lda The stride between columns of A in terms of elements of
+ *             type S.
+ *  @param B The target matrix buffer.
+ *  @param ldb The stride between columns of B in terms of elements of
+ *             type T.
+ *  @param func The functor to apply. Must be device-invocable.
+ */
+template <typename S, typename T, typename SizeT, typename FunctorT>
+void CombineImpl(
+    SizeT m, SizeT n,
+    S const* A, SizeT lda,
+    T * B, SizeT ldb,
+    FunctorT func,
+    SyncInfo<Device::GPU> const& sync_info)
+{
+
+    if (m == TypeTraits<SizeT>::Zero() || n == TypeTraits<SizeT>::Zero())
+    {
+        // Nothing to do
+        return;
+    }
+
+    constexpr int TILE_DIM = gpu::Default2DTileSize();
+    constexpr int BLK_COLS = 8;
+
+    static_assert(TILE_DIM % BLK_COLS == 0,
+                  "Incompatible TILE_DIM, BLK_COLS.");
+
+    dim3 blks((m + TILE_DIM - 1) / TILE_DIM,
+              (n + TILE_DIM - 1) / TILE_DIM,
```

Comment:
Technically this is only true in CUDA. In that case, we should consider that case both here and in other kernels. Since this applies to many other kernels in Hydrogen, I'm going to open an issue to track that work elsewhere and resolve it with a future PR so as not to extend the scope of this one.

Reply:
Ah, good point. I suppose this isn't trivial either since it requires reworking the kernels.

Reply:
Opened issue #110.